File: tests/cuda/cpp/test_signal_filter_design.cpp
// tests/cuda/test_signal_filter_design.cpp
#include "insight/insight.h"
#include <cmath>
#include <complex>
#include <gtest/gtest.h>
#include <vector>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace ins;

namespace {

class SignalFilterDesignTestGPU : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    ins::init({"cpu", "iluvatar"});
    try {
      set_device(GPUPlace(0));
    // Iluvatar: signal API uses F64
    GTEST_SKIP() << "Iluvatar: signal API uses F64 (hardware limit)";
    } catch (...) {
      GTEST_SKIP() << "GPU not available";
    }
  }
};

template <typename T>
void expect_float_equal(const Array &arr, const std::vector<T> &expected,
                        T tol = 1e-6) {
  ASSERT_EQ(arr.numel(), static_cast<int64_t>(expected.size()));
  const T *data = arr.data<T>();
  for (int64_t i = 0; i < arr.numel(); ++i) {
    EXPECT_NEAR(data[i], expected[i], tol) << "mismatch at index " << i;
  }
}

// ========== kaiser_beta (scalar, no GPU/CPU difference) ==========

TEST_F(SignalFilterDesignTestGPU, KaiserBetaLow) {
  EXPECT_DOUBLE_EQ(signal::kaiser_beta(0.0), 0.0);
  EXPECT_DOUBLE_EQ(signal::kaiser_beta(21.0), 0.0);
}

TEST_F(SignalFilterDesignTestGPU, KaiserBetaMid) {
  double a = 40.0;
  double expected = 0.5842 * std::pow(a - 21.0, 0.4) + 0.07886 * (a - 21.0);
  EXPECT_NEAR(signal::kaiser_beta(a), expected, 1e-10);
}

TEST_F(SignalFilterDesignTestGPU, KaiserBetaHigh) {
  double a = 60.0;
  double expected = 0.1102 * (a - 8.7);
  EXPECT_NEAR(signal::kaiser_beta(a), expected, 1e-10);
}

TEST_F(SignalFilterDesignTestGPU, KaiserBetaBoundary) {
  double a = 50.0;
  double beta = signal::kaiser_beta(a);
  EXPECT_GT(beta, 4.0);
  EXPECT_LT(beta, 6.0);
}

// ========== kaiser_atten (scalar, no GPU/CPU difference) ==========

TEST_F(SignalFilterDesignTestGPU, KaiserAttenBasic) {
  double atten = signal::kaiser_atten(101, 0.1);
  double expected = 2.285 * 100.0 * M_PI * 0.1 + 7.95;
  EXPECT_NEAR(atten, expected, 1e-10);
}

TEST_F(SignalFilterDesignTestGPU, KaiserAttenNarrow) {
  double atten_wide = signal::kaiser_atten(101, 0.2);
  double atten_narrow = signal::kaiser_atten(101, 0.05);
  EXPECT_GT(atten_wide, atten_narrow);
}

// ========== firwin ==========

TEST_F(SignalFilterDesignTestGPU, FirwinLowpassBasic) {
  // firwin returns CPU array (composite ops use CPU data internally)
  Array h = signal::firwin(11, {0.3}, "hamming", "lowpass", true);
  Array h_gpu = h.to(GPUPlace(0));
  Array h_cpu = h_gpu.to(CPUPlace());
  EXPECT_EQ(h_cpu.numel(), 11);
  const double *d = h_cpu.data<double>();
  for (int i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], d[10 - i], 1e-10) << "symmetry failed at index " << i;
  }
}

TEST_F(SignalFilterDesignTestGPU, FirwinLowpassDCGain) {
  Array h = signal::firwin(21, {0.25}, "hamming", "lowpass", true);
  Array h_gpu = h.to(GPUPlace(0));
  Array dc_gpu = sum(h_gpu);
  Array dc_cpu = dc_gpu.to(CPUPlace());
  EXPECT_NEAR(dc_cpu.data<double>()[0], 1.0, 1e-6);
}

TEST_F(SignalFilterDesignTestGPU, FirwinHighpassBasic) {
  Array h = signal::firwin(11, {0.3}, "hamming", "highpass", true);
  Array h_gpu = h.to(GPUPlace(0));
  Array h_cpu = h_gpu.to(CPUPlace());
  EXPECT_EQ(h_cpu.numel(), 11);
  const double *d = h_cpu.data<double>();
  for (int i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], d[10 - i], 1e-10);
  }
}

TEST_F(SignalFilterDesignTestGPU, FirwinHighpassNyquistGain) {
  Array h = signal::firwin(21, {0.25}, "hamming", "highpass", true);
  const double *d = h.data<double>();
  double nyquist_gain = 0.0;
  for (int i = 0; i < 21; ++i) {
    nyquist_gain += d[i] * (i % 2 == 0 ? 1.0 : -1.0);
  }
  EXPECT_NEAR(nyquist_gain, 1.0, 1e-6);
}

TEST_F(SignalFilterDesignTestGPU, FirwinBandpass) {
  Array h = signal::firwin(21, {0.2, 0.5}, "hamming", "bandpass", false);
  Array h_gpu = h.to(GPUPlace(0));
  Array dc_gpu = sum(h_gpu);
  Array dc_cpu = dc_gpu.to(CPUPlace());
  EXPECT_NEAR(dc_cpu.data<double>()[0], 0.0, 0.01);
}

TEST_F(SignalFilterDesignTestGPU, FirwinBandstop) {
  Array h = signal::firwin(21, {0.2, 0.5}, "hamming", "bandstop", false);
  EXPECT_EQ(h.numel(), 21);
}

TEST_F(SignalFilterDesignTestGPU, FirwinDifferentWindows) {
  Array h1 = signal::firwin(11, {0.3}, "hann", "lowpass", false);
  Array h2 = signal::firwin(11, {0.3}, "blackman", "lowpass", false);
  Array h3 = signal::firwin(11, {0.3}, "kaiser", "lowpass", false);
  EXPECT_EQ(h1.numel(), 11);
  EXPECT_EQ(h2.numel(), 11);
  EXPECT_EQ(h3.numel(), 11);
  EXPECT_NE(h1.data<double>()[0], h2.data<double>()[0]);
}

TEST_F(SignalFilterDesignTestGPU, FirwinUnscaled) {
  Array h = signal::firwin(21, {0.25}, "hamming", "lowpass", false);
  Array dc = sum(h);
  double dc_val = dc.data<double>()[0];
  EXPECT_NE(dc_val, 1.0);
}

// ========== firwin2 ==========

TEST_F(SignalFilterDesignTestGPU, Firwin2Lowpass) {
  std::vector<double> freq = {0.0, 0.5, 0.5, 1.0};
  std::vector<double> gain = {1.0, 1.0, 0.0, 0.0};
  Array h = signal::firwin2(15, freq, gain, 0, "hamming", false);
  Array h_gpu = h.to(GPUPlace(0));
  Array dc_gpu = sum(h_gpu);
  Array dc_cpu = dc_gpu.to(CPUPlace());
  EXPECT_EQ(h.numel(), 15);
  EXPECT_NEAR(dc_cpu.data<double>()[0], 1.0, 0.1);
}

TEST_F(SignalFilterDesignTestGPU, Firwin2Highpass) {
  std::vector<double> freq = {0.0, 0.5, 0.5, 1.0};
  std::vector<double> gain = {0.0, 0.0, 1.0, 1.0};
  Array h = signal::firwin2(15, freq, gain, 0, "hamming", false);
  EXPECT_EQ(h.numel(), 15);
}

TEST_F(SignalFilterDesignTestGPU, Firwin2Bandpass) {
  std::vector<double> freq = {0.0, 0.2, 0.2, 0.5, 0.5, 1.0};
  std::vector<double> gain = {0.0, 0.0, 1.0, 1.0, 0.0, 0.0};
  Array h = signal::firwin2(21, freq, gain, 0, "hamming", false);
  EXPECT_EQ(h.numel(), 21);
}

TEST_F(SignalFilterDesignTestGPU, Firwin2DifferentLengths) {
  std::vector<double> freq = {0.0, 0.5, 1.0};
  std::vector<double> gain = {1.0, 1.0, 0.0};
  Array h_short = signal::firwin2(7, freq, gain);
  Array h_long = signal::firwin2(31, freq, gain);
  EXPECT_EQ(h_short.numel(), 7);
  EXPECT_EQ(h_long.numel(), 31);
}

// ========== cmplx_sort ==========

TEST_F(SignalFilterDesignTestGPU, CmplxSortReal) {
  std::vector<double> data = {3.0, -1.0, 2.0, -4.0};
  Array arr = to_array(data);
  Array sorted = signal::cmplx_sort(arr);
  const double *d = sorted.data<double>();
  EXPECT_NEAR(d[0], -1.0, 1e-10);
  EXPECT_NEAR(d[1], 2.0, 1e-10);
  EXPECT_NEAR(d[2], 3.0, 1e-10);
  EXPECT_NEAR(d[3], -4.0, 1e-10);
}

TEST_F(SignalFilterDesignTestGPU, CmplxSortComplex) {
  std::vector<std::complex<double>> data = {{0, 5}, {1, 0}, {0, 3}, {4, 0}};
  Array arr = to_array(data, DType::C64);
  Array sorted = signal::cmplx_sort(arr);
  const std::complex<double> *d =
      reinterpret_cast<const std::complex<double> *>(sorted.data<char>());
  EXPECT_NEAR(std::abs(d[0]), 1.0, 1e-10);
  EXPECT_NEAR(std::abs(d[1]), 3.0, 1e-10);
  EXPECT_NEAR(std::abs(d[2]), 4.0, 1e-10);
  EXPECT_NEAR(std::abs(d[3]), 5.0, 1e-10);
}

TEST_F(SignalFilterDesignTestGPU, CmplxSortSingle) {
  std::vector<double> data = {42.0};
  Array arr = to_array(data);
  Array sorted = signal::cmplx_sort(arr);
  EXPECT_NEAR(sorted.data<double>()[0], 42.0, 1e-10);
}

TEST_F(SignalFilterDesignTestGPU, CmplxSortZeros) {
  std::vector<double> data = {0.0, 0.0, 0.0};
  Array arr = to_array(data);
  Array sorted = signal::cmplx_sort(arr);
  EXPECT_NEAR(sorted.data<double>()[0], 0.0, 1e-10);
  EXPECT_NEAR(sorted.data<double>()[1], 0.0, 1e-10);
  EXPECT_NEAR(sorted.data<double>()[2], 0.0, 1e-10);
}

} // namespace

File: tests/cuda/python/test_signal_filter_design.py
"""Signal Filter Design CUDA binding tests.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python \
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
    python -m pytest tests/cuda/python/test_signal_filter_design.py -v
"""

import sys
import os
import pytest
import math

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)
try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)


class TestKaiserBeta:
    """Kaiser beta parameter — tests 1-4."""

    def test_kaiser_beta_low(self):
        # a <= 21 -> beta = 0
        assert ins.signal.kaiser_beta(0.0) == 0.0
        assert ins.signal.kaiser_beta(21.0) == 0.0

    def test_kaiser_beta_mid(self):
        a = 40.0
        expected = 0.5842 * (a - 21.0) ** 0.4 + 0.07886 * (a - 21.0)
        assert abs(ins.signal.kaiser_beta(a) - expected) < 1e-10

    def test_kaiser_beta_high(self):
        a = 60.0
        expected = 0.1102 * (a - 8.7)
        assert abs(ins.signal.kaiser_beta(a) - expected) < 1e-10

    def test_kaiser_beta_boundary(self):
        beta = ins.signal.kaiser_beta(50.0)
        assert beta > 4.0
        assert beta < 6.0


class TestKaiserAtten:
    """Kaiser attenuation — test 5."""

    def test_kaiser_atten_basic(self):
        atten = ins.signal.kaiser_atten(101, 0.1)
        expected = 2.285 * 100.0 * math.pi * 0.1 + 7.95
        assert abs(atten - expected) < 1e-10


class TestFirwin:
    """FIR filter design — tests 6-12."""

    def test_firwin_lowpass_basic(self):
        h = ins.signal.firwin(11, [0.3], window="hamming", pass_zero="lowpass")
        assert h.numel == 11
        d = h.numpy()
        # Symmetric (Type I: odd numtaps)
        for i in range(5):
            assert abs(d[i] - d[10 - i]) < 1e-10

    def test_firwin_lowpass_dc_gain(self):
        h = ins.signal.firwin(21, [0.25], window="hamming", pass_zero="lowpass")
        dc = float(np.sum(h.numpy()))
        assert abs(dc - 1.0) < 1e-6

    def test_firwin_highpass_basic(self):
        h = ins.signal.firwin(11, [0.3], window="hamming", pass_zero="highpass")
        assert h.numel == 11
        d = h.numpy()
        for i in range(5):
            assert abs(d[i] - d[10 - i]) < 1e-10

    def test_firwin_highpass_nyquist_gain(self):
        h = ins.signal.firwin(21, [0.25], window="hamming", pass_zero="highpass")
        d = h.numpy()
        nyquist_gain = sum(d[i] * (1.0 if i % 2 == 0 else -1.0) for i in range(21))
        assert abs(nyquist_gain - 1.0) < 1e-6

    def test_firwin_bandpass(self):
        h = ins.signal.firwin(21, [0.2, 0.5], window="hamming", pass_zero="bandpass")
        assert h.numel == 21
        dc = float(np.sum(h.numpy()))
        assert abs(dc) < 0.01  # DC gain ~0 for bandpass

    def test_firwin_different_windows(self):
        h1 = ins.signal.firwin(11, [0.3], window="hann", pass_zero="lowpass")
        h2 = ins.signal.firwin(11, [0.3], window="blackman", pass_zero="lowpass")
        assert h1.numel == 11
        assert h2.numel == 11
        assert h1.numpy()[0] != h2.numpy()[0]

    def test_firwin_unscaled(self):
        h = ins.signal.firwin(21, [0.25], window="hamming", pass_zero="lowpass", scale=False)
        dc = float(np.sum(h.numpy()))
        assert dc != 1.0


class TestFirwin2:
    """FIR filter design (arbitrary response) — tests 13-14."""

    def test_firwin2_lowpass(self):
        freq = [0.0, 0.5, 0.5, 1.0]
        gain = [1.0, 1.0, 0.0, 0.0]
        h = ins.signal.firwin2(15, freq, gain)
        assert h.numel == 15
        dc = float(np.sum(h.numpy()))
        assert abs(dc - 1.0) < 0.1

    def test_firwin2_different_lengths(self):
        freq = [0.0, 0.5, 1.0]
        gain = [1.0, 1.0, 0.0]
        h_short = ins.signal.firwin2(7, freq, gain)
        h_long = ins.signal.firwin2(31, freq, gain)
        assert h_short.numel == 7
        assert h_long.numel == 31


class TestCmplxSort:
    """Complex sort — tests 15-16."""

    def test_cmplx_sort_real(self):
        a = ins.from_numpy(np.array([3.0, -1.0, 2.0, -4.0], dtype=np.float64))
        result = ins.signal.cmplx_sort(a)
        # cmplx_sort returns (sorted, indices)
        if isinstance(result, tuple):
            sorted_arr = result[0]
        else:
            sorted_arr = result
        d = sorted_arr.numpy()
        assert abs(d[0] - (-1.0)) < 1e-10
        assert abs(d[1] - 2.0) < 1e-10
        assert abs(d[2] - 3.0) < 1e-10
        assert abs(d[3] - (-4.0)) < 1e-10

    def test_cmplx_sort_single(self):
        a = ins.from_numpy(np.array([42.0], dtype=np.float64))
        result = ins.signal.cmplx_sort(a)
        if isinstance(result, tuple):
            sorted_arr = result[0]
        else:
            sorted_arr = result
        assert abs(sorted_arr.numpy()[0] - 42.0) < 1e-10


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
